README.md
iTunes XML parser for PHP
==========================

#### A class to extract info, tracks and playlists from an iTunes XML file.

Original code copyright (c) 2003 by [Robert A. Wallis](https://github.com/robert-wallis).
Dicked around with in 2005 by [Peter Minarik](http://www.wirsindecht.org/),
and in 2013-2021 by [Conan Theobald](https://github.com/shuckster).

LGPL2.1 licensed: See [LICENSE](LICENSE)

## About

A simple PHP class that will read an iTunes XML file and convert the info,
tracks and playlists contained within into an array of objects.

Tracks can be matched to playlist-items by running the `#processPlaylists()`
method after opening your XML file.

Supports sorting fields by string, number, and date.

## Instructions

See [example.php](example.php) for a basic implementation.

```php
require_once 'iTunesXMLparser.class.php';

$xml_path = 'iTunes playlist export.xml';

$itunes = new iTunesXMLParser();
$itunes->sort_field = 'Track ID';
$itunes->sort_direction = 'ascending';
$itunes->open( $xml_path );

/*
  "$itunes->data" is now available. print_r to see what's inside:
    print_r( $itunes->data );
*/

```

For conversion to JSON, I recommend using the `jsbeautifier.org`
[PHP port](https://github.com/einars/js-beautify/tree/attic-php/php).

EDIT - No need for that [now](https://www.php.net/manual/en/function.json-encode.php). :)

## 2021 update

Before releasing 1.6 I discovered a fork by @dajoho that has a `composer.json`:

*   https://github.com/dajoho/itunes-php

This version had also already fixed the [sort() bug](https://github.com/shuckster/iTunes-XML-parser-for-PHP/commit/1d02fb4c0b93309712f8bcf1733d2746cdb90737) [2 years ago](https://github.com/dajoho/itunes-php/commit/eb28d8a3873607dc57aa412f6cffcafa58173f57)! :)

## Contributors

*   [@nickbe](https://github.com/nickbe) - Bug report
*   [@dajoho](https://github.com/dajoho) - Composer compatible fork
*   [@SidRoberts](https://github.com/SidRoberts) - Code improvement


## Credits

Based on work by:

*   Robert Wallis
*   http://www.wirsindecht.org/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project does *NOT* adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). But it might do after 1.6. Maybe. ;)

## [Unreleased]

## [1.6] - 2021-03-02
### Fixed
- Fix sort() accessors

## [1.5] - 2013-07-07
### Updated
- Simplify parseDict, API changes

## [1.4] - 2013-06-23
### Added
- Parse info and playlists

## [1.1 - 1.3] - 2013-03-22
### Refactoring
- New example
- Delete old/deprecated stuff
- Now a class, improved sort-method
- Type-cast integers and booleans


example.php
<?php
/*
  Example for using iTunes XML parser for PHP.
  Copyright (C) 2013-2021 Conan Theobald [http://github.com/shuckster]
  Source repository:
  - https://github.com/shuckster/iTunes-XML-parser-for-PHP

  based on:

    Copyright (C) 2005 Peter Minarik [http://www.wirsindecht.org]
    version: 1.00
    based on:

    iTunes XML PhP Parser
    Copyright (C) 2003 Robert A. Wallis [https://github.com/robert-wallis]
    version: 1.00

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE
  as published by the Free Software Foundation; either version 2.1
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free
  Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
  MA 02111-1307, USA

  iTunes is a product by Apple Computer, Inc.
  http://www.apple.com/

*/

require_once 'iTunesXMLparser.class.php';

$xml_path = 'iTunes playlist export.xml';

$itunes = new iTunesXMLParser();
$itunes->sort_field = 'Track ID';
$itunes->sort_direction = 'ascending';
$itunes->open( $xml_path );

/*
  "$itunes->data" is now available. print_r to see what's inside:
    print_r( $itunes->data );
*/

// Find only videos (kind of useless, since the example XML only
// has video in it ;)
$video = array();
$tracks = (array) $itunes->data[ 'Tracks' ];
foreach ( $tracks as $track ) {
	if ( isset( $track->{ 'Has Video' } ) && $track->{ 'Has Video' } && isset( $track->{ 'Location' } ) ) {
		$video[] = $track;
	}
}

print_r( $video );
exit;

/*
 * Old JSON output example using JSBeautify.
 * These days we can use json_encode()
 */

// JSBeautify and options
// Download from: https://github.com/einars/js-beautify/tree/attic-php/php

require_once 'jsbeautifier.php';

$jsb = new JSBeautifier();
$jsb_opts = new BeautifierOptions();
$jsb_opts->indent_size = 1;
$jsb_opts->indent_with_tabs = true;

header( 'Content-type: application/json; charset=utf-8' );
echo $jsb->beautify( json_encode( $video ), $jsb_opts );
exit;

?>


example_simple.php
<?php
/*
  Copyright (C) 2003 Robert A. Wallis http://codetriangle.com

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
*/

// must do this, include itunes_xml_parser_php5.php if your server is php v5
include "itunes_xml_parser.php";

// get songs from the xml file
$songs = iTunesXmlParser("techno_playlist.xml");

// if it worked
if ($songs)
{
	// create a table
	$output = '<table cellpadding="0" cellspacing="0" border="0" width="100%">';
	
	// fill the first row with headings
	$output .= '<tr>';
	$output .= '<td><font size="1">name</font></td>';
	$output .= '<td><font size="1">artist</font></td>';
	$output .= '<td><font size="1">album</font></td>';
	$output .= '<td><font size="1">rating</font></td>';
	$output .= '</tr>';
		
	// loop through the songs in the array and get 4 fields that I want to see
	foreach ($songs as $song)
	{
		$output .= '<tr>';
		$output .= '<td><font size="2">'.$song["Name"].'</font></td>';
		$output .= '<td><font size="2">'.$song["Artist"].'</font></td>';
		$output .= '<td><font size="2">'.$song["Album"].'</font></td>';
		$output .= '<td><font size="2">'.$song["Rating"].'</font></td>';
		$output .= '</tr>';
	}
	
	// end the table
	$output .= '</table>';
	
	// show my new table
	print ($output);
}


example_advanced.php
<?php
/*
  Copyright (C) 2003 Robert A. Wallis http://codetriangle.com

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
*/

include "itunes_xml_parser.php";

// filename is the xml file, cat prefix is something different for each time you call the function
// ie. DrawPlaylist("techno_playlist.xml", "techno-");
function DrawPlaylist($filename, $cat_prefix)
{
	$output = NULL;
	
	// get "techno-f" and "tehcno-d" vars if they exsist
	// f stands for field, and d stants for direction
	$field = isset($_REQUEST[$cat_prefix."f"])?$_REQUEST[$cat_prefix."f"]:NULL;
	$direction = isset($_REQUEST[$cat_prefix."d"])?$_REQUEST[$cat_prefix."d"]:NULL;
	
	// load up the list of songs and sort
	$songs = iTunesXmlParser($filename, $field, $direction);
	
	// if the load was successful
	if ($songs)
	{
		// create an anchor so when the list is sorted, it comes back to where you are now
		$output .= '<a name="'.$cat_prefix.'"></a>';
		
		// start a table
		$output .= '<table cellpadding="0" cellspacing="0" border="0" width="100%">';
		$output .= '<tr>';
		// for each field, create a link that will toggle up or down sorting direction
		// also add the anchor to the end of the link so when the page refreshes these
		// fields stay in the viewing area
		$output .= '<td><font size="1"><a href="example_advanced.php?'.$cat_prefix.'f=Name&'.$cat_prefix.'d='.(($field=="Name")?(($direction=="up")?"down":"up"):"down").'#'.$cat_prefix.'">name</a></font></td>';
		$output .= '<td><font size="1"><a href="example_advanced.php?'.$cat_prefix.'f=Artist&'.$cat_prefix.'d='.(($field=="Artist")?(($direction=="up")?"down":"up"):"down").'#'.$cat_prefix.'">artist</a></font></td>';
		$output .= '<td><font size="1"><a href="example_advanced.php?'.$cat_prefix.'f=Album&'.$cat_prefix.'d='.(($field=="Album")?(($direction=="up")?"down":"up"):"down").'#'.$cat_prefix.'">album</a></font></td>';
		$output .= '<td><font size="1"><a href="example_advanced.php?'.$cat_prefix.'f=Rating&'.$cat_prefix.'d='.(($field=="Rating")?(($direction=="up")?"down":"up"):"down").'#'.$cat_prefix.'">rating</a></font></td>';
		$output .= '</tr>';
		
		// now actually get the songs
		foreach ($songs as $song)
		{
			// display the field if the field is set
			$output .= '<tr>';
			$output .= '<td><font size="2">'.(isset($song["Name"])?$song["Name"]:NULL).'</font></td>';
			$output .= '<td><font size="2">'.(isset($song["Artist"])?$song["Artist"]:NULL).'</font></td>';
			$output .= '<td><font size="2">'.(isset($song["Album"])?$song["Album"]:NULL).'</font></td>';
			$output .= '<td><font size="2">'.(isset($song["Rating"])?$song["Rating"]:NULL).'</font></td>';
			$output .= '</tr>';
		}
		
		// end the table
		$output .= '</table>';
	}
	
	// show our results
	print ($output);
}


// *****************
// now for each playlist you can have a separate table

DrawPlaylist("techno_playlist.xml", "tech-");


itunes_xml_parser.php
<?php
/* 
  iTunes XML PhP Parser
  Copyright (C) 2003 Robert A. Wallis [http://codetriangle.com/]
  version: 1.00

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE
  as published by the Free Software Foundation; either version 2.1
  of the License, or (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.
    
  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free
  Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
  MA 02111-1307, USA 
  
  iTunes is a product by Apple Computer, Inc.
  http://www.apple.com/
  
*/

// pass the $filename of the xml file you want to parse
// [optional] $sort_field = can be set to "Album", "Rating", etc...  left alone will not sort the list
// [optional] $sort_direction = "up" or "down"
//
// the function will return an array of songs
// each song is an array of variables that makes sense
/* example output
Array ( 
	[0] => Array 
		( 
			[Track ID] => 34
			[Name] => depeche mode gameboy megamix!
			[Artist] => nullsleep
			[Composer] => nullsleep
			[Album] => www.nullsleep.com
			[Genre] => Electronic
			[Kind] => MPEG audio file
			[Size] => 13901952
			[Total Time] => 868780
			[Year] => 2002
			[Date Modified] => 2002-12-17T10:24:18Z
			[Date Added] => 2003-11-11T06:38:24Z
			[Bit Rate] => 128
			[Sample Rate] => 44100
			[Comments] => www.8bitpeoples.com 1. enjoy the silence 2. photographic 3. new life 4. everything counts
			[Play Count] => 3
			[Play Date] => -1142117754
			[Play Date UTC] => 2003-11-28T15:32:22Z
			[Rating] => 80
			[Normalization] => 976
			[Location] => file://localhost/C:/media/nullsleep/www.nullsleep.com/depeche%20mode%20gameboy%20megamix!.mp3/
			[File Folder Count] => 4
			[Library Folder Count] => 1
		)
    [1] => Array
		(
			[Track ID] => 65
			[Name] => Daftendirekt
			[Artist] => Daft Punk
			[Album] => Homework
			[Genre] => Electronic
			[Kind] => MPEG audio file
			[Size] => 4098756
			[Total Time] => 164649
			[Track Number] => 1
			[Date Modified] => 2003-11-28T19:45:05Z
			[Date Added] => 2003-11-11T06:38:25Z
			[Bit Rate] => 192
			[Sample Rate] => 44100
			[Play Count] => 5
			[Play Date] => -1142119790
			[Play Date UTC] => 2003-11-28T14:58:26Z
			[Rating] => 80
			[Normalization] => 1414
			[Location] => file://localhost/C:/media/Daft%20Punk/Homework/01%20Daftendirekt.MP3/
			[File Folder Count] => 4
			[Library Folder Count] => 1
		)
}
*/
function iTunesXmlParser($filename, $sort_field=NULL, $sort_direction="up")
{
	// save the input in global variables for the sort function
	global $g_ITX_field, $g_ITX_direction;
	$g_ITX_field = $sort_field;
	$g_ITX_direction = $sort_direction;
	
	// init main variables
	$songs = array();
	$xml = NULL; // parsed XML
	
	// read the file into $xml first
	ob_start();
		readfile($filename);
		$xml = ob_get_contents();
	ob_end_clean();
	
	// open the xml document in the DOM
	if (!$xml || !$dom = domxml_open_mem($xml))
		die("Could not parse iTunes XML file: ".$filename);
	
	// get the root element
	$root = $dom->document_element();
	
	// yeah "dict" means everything, playlist, and song that makes sense... NOT
	// find the first "dict"
	$children = $root->child_nodes();
	foreach ($children as $child)
	{
		if ($child->node_name()=="dict")
		{
			$root = $child;
			break;
		}
	}

	// do that again, and find the second inner dict
	$children = $root->child_nodes();
	foreach ($children as $child)
	{
		if ($child->node_name()=="dict")
		{
			$root = $child;
			break;
		}
	}
		
	// now go through all the child elements
	$children = $root->child_nodes();
	foreach ($children as $child)
	{
		// all the sub dicts from here on should be songs
		if ($child->node_name()=="dict")
		{
			$song = NULL;
			
			// get all the elements
			$elements = $child->child_nodes();
			for ($i = 0; $i<count($elements); $i++)
			{
				// alright whomever wrote this xml file was smoking something serious
				// in normal XML documents we would do:
				//  <artist>Daft Punk</artist>
				// but in Apple iTunes bong land we do:
				//  <key>Artist</key><string>Daft Punk</string>
				
				if ($elements[$i]->node_name()=="key")
				{
					// so I'm just going to expect that i++ (<string>, <int>, etc...) is always going to be there,
					//  if the key's name is <key>
					//  instead of doing some error checking here to make sure there are matching values to keys
					$key = $elements[$i]->get_content();
					$i++;
					$value = $elements[$i]->get_content();
					$song[$key]=$value;
				}
			}
			
			// save the song
			if ($song)
				$songs[] = $song;
		}
	}
	
	// now sort the songs
	// $sort_field=NULL, $sort_direction="up"
	if ($sort_field)
	{
		uasort($songs, "iTunesXmlSongSort");
	}

	return $songs;
}

$g_ITX_field = NULL;
$g_ITX_direction = NULL;
// to be used with the uasort() array function in PHP
function iTunesXmlSongSort($left, $right)
{
	global $g_ITX_field, $g_ITX_direction;
	
	// return the strcmp() of the two fields
	if (isset($left[$g_ITX_field])&&isset($right[$g_ITX_field]))
	{
		if (strcasecmp($g_ITX_direction, "up"))
			return strcasecmp($left[$g_ITX_field],$right[$g_ITX_field]);
		else
			return strcasecmp($right[$g_ITX_field],$left[$g_ITX_field]);	
	}
	elseif (isset($left[$g_ITX_field]))
		return -1;
	else
		return 1;
}


itunes_xml_parser_php5.php
<?php
/*
  iTunes XML PhP Parser for PHP 5
  Copyright (C) 2013 Conan Theobald [http://github.com/shuckster]
  version: 1.2
  	Changes:
  		* 1.2: Now a class, improved sort-method
  		* 1.1: Type-cast integers and booleans

  based on:

  Copyright (C) 2005 Peter Minarik [http://www.wirsindecht.org]
  version: 1.00
  based on:

  iTunes XML PhP Parser
  Copyright (C) 2003 Robert A. Wallis [http://codetriangle.com/]
  version: 1.00

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE
  as published by the Free Software Foundation; either version 2.1
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free
  Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
  MA 02111-1307, USA

  iTunes is a product by Apple Computer, Inc.
  http://www.apple.com/

*/

// pass the $filename of the xml file you want to parse
// [optional] $sort_field = can be set to 'Album', 'Rating', etc...  left alone will not sort the list
// [optional] $sort_direction = 'up' or 'down'
//
// the function will return an array of songs
// each song is an array of variables that makes sense
/* example output
Array (
	[0] => Array
		(
			[Track ID] => 34
			[Name] => depeche mode gameboy megamix!
			[Artist] => nullsleep
			[Composer] => nullsleep
			[Album] => www.nullsleep.com
			[Genre] => Electronic
			[Kind] => MPEG audio file
			[Size] => 13901952
			[Total Time] => 868780
			[Year] => 2002
			[Date Modified] => 2002-12-17T10:24:18Z
			[Date Added] => 2003-11-11T06:38:24Z
			[Bit Rate] => 128
			[Sample Rate] => 44100
			[Comments] => www.8bitpeoples.com 1. enjoy the silence 2. photographic 3. new life 4. everything counts
			[Play Count] => 3
			[Play Date] => -1142117754
			[Play Date UTC] => 2003-11-28T15:32:22Z
			[Rating] => 80
			[Normalization] => 976
			[Location] => file://localhost/C:/media/nullsleep/www.nullsleep.com/depeche%20mode%20gameboy%20megamix!.mp3/
			[File Folder Count] => 4
			[Library Folder Count] => 1
		)
    [1] => Array
		(
			[Track ID] => 65
			[Name] => Daftendirekt
			[Artist] => Daft Punk
			[Album] => Homework
			[Genre] => Electronic
			[Kind] => MPEG audio file
			[Size] => 4098756
			[Total Time] => 164649
			[Track Number] => 1
			[Date Modified] => 2003-11-28T19:45:05Z
			[Date Added] => 2003-11-11T06:38:25Z
			[Bit Rate] => 192
			[Sample Rate] => 44100
			[Play Count] => 5
			[Play Date] => -1142119790
			[Play Date UTC] => 2003-11-28T14:58:26Z
			[Rating] => 80
			[Normalization] => 1414
			[Location] => file://localhost/C:/media/Daft%20Punk/Homework/01%20Daftendirekt.MP3/
			[File Folder Count] => 4
			[Library Folder Count] => 1
		)
}
*/

class iTunesXMLParser {

	public $file_name = '';
	public $tracks = array();

	public $sort_field = NULL;
	public $sort_direction = 'ascending';

	public function open( $path ) {

		if ( !file_exists( $path ) ) {
			die( 'iTunes XML file not found: ' . $path );
		}

		// Open the XML document in the DOM
		$dom = new DomDocument();
		if ( !$dom->load( $path ) ) {
			die( 'Could not parse iTunes XML file: ' . $path );
		}

		// Get the root element <plist>
		$plist_node = $dom->documentElement;
		$version_node = NULL;
		$tracks_node = array();

		// First <dict> contains version-info + tracks-node
		foreach ( $plist_node->childNodes as $child ) {
			if ( 'dict' === $child->nodeName ) {
				$version_node = $child;
				break;
			}
		}

		// <dict> in the version-node contains tracks
		foreach ( $version_node->childNodes as $child ) {
			if ( 'dict' === $child->nodeName ) {
				$tracks_node = $child;
				break;
			}
		}

		// Loop through the tracks
		foreach ( $tracks_node->childNodes as $child ) {

			// all the sub dicts from here on should be songs
			if ( 'dict' === $child->nodeName ) {
				$track = NULL;

				// Get track properties
				$properties = $child->childNodes;
				for ( $prop_index = 0, $prop_length = $properties->length; $prop_index < $prop_length; $prop_index++ ) {

					$prop = $properties->item( $prop_index );

					// Ordering is important in plist files; key -> value is part of the convention

					if ( 'key' === $prop->nodeName ) {

						// <key> $prop_key </key>
						$prop_key = $prop->textContent;

						// <value> $prop_value </value>
						$prop = $properties->item( ++$prop_index );

						switch ( $prop->nodeName ) {
							case 'true':
							case 'false':
								$prop_value = 'true' === $prop->nodeName;
							break;

							case 'integer':
								$prop_value = (int) $prop->textContent;
							break;

							default:
								$prop_value = $prop->textContent;
						}

						$track[ $prop_key ] = $prop_value;
					}
				}

				// Save the track
				if ( $track ) {
					$tracks[] = $track;
				}

			}

		}

		// Sort the tracks
		if ( $this->sort_field ) {
			uasort( $tracks, array( $this, 'sort' ) );
		}

		// Fell-through: Set public vars to successful parse-data
		$this->file_name = $path;
		$this->tracks = $tracks;

		return $tracks;

	}

	// To be used with the uasort() array function
	protected function sort( $left, $right ) {

		$field = $this->sort_field;
		$direction = $this->sort_direction;

		// Return the strcmp() of the two fields
		if ( isset( $left[ $field ] ) && isset( $right[ $field ] ) ) {

			$left = $left[ $field ];
			$right = $right[ $field ];

			switch ( gettype( $left ) ) {

				case 'boolean':
					$left = (int) $left;
					$right = (int) $right;

				case 'integer':
				case 'double':
					if ( 'descending' === $direction ) {
						return $left === $right ? 0 : ( $left > $right ? -1 : 1 );
					}
					else {
						return $left === $right ? 0 : ( $right > $left ? -1 : 1 );
					}
				break;

				default:
					if ( 'descending' === $direction ) {
						return strcasecmp( $left, $right );
					}
					else {
						return strcasecmp( $right, $left );
					}

			}

		}
		elseif ( isset( $left[ $field ] ) ) {
			return -1;
		}
		else {
			return 1;
		}

	}

}


iTunesXMLparser.class.php
<?php
/*
  iTunes XML parser for PHP
  Copyright (C) 2013-2021 Conan Theobald [http://github.com/shuckster]
  Version: 1.6

  Source repository:
  - https://github.com/shuckster/iTunes-XML-parser-for-PHP

  Based on:

    Copyright (C) 2005 Peter Minarik [http://www.wirsindecht.org]
    version: 1.00
    based on:

    iTunes XML PhP Parser
    Copyright (C) 2003 Robert A. Wallis [https://github.com/robert-wallis]
    version: 1.00

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU LESSER GENERAL PUBLIC LICENSE
  as published by the Free Software Foundation; either version 2.1
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free
  Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
  MA 02111-1307, USA

  iTunes is a product by Apple Computer, Inc.
  http://www.apple.com/

*/

class iTunesXMLParser {

  public $file_name = '';
  public $data = NULL;

  public $sort_field = NULL;
  public $sort_direction = 'ascending';

  //
  // INTERFACE
  //
  public function parse( $source ) {
    return $this->openFileOrSource( NULL, $source );
  }

  public function open( $file ) {
    return $this->openFileOrSource( $file );
  }

  public function processPlaylists() {

    if ( NULL === $this->data || !isset( $this->data[ 'Playlists' ] ) ) {
      die( 'No data to work with' );
    }

    $tracks = (array) $this->data[ 'Tracks' ];

    foreach( $this->data[ 'Playlists' ] as &$playlist ) {
      $new_items = array();

      foreach ( $playlist->{ 'Playlist Items' } as $item ) {
        $track_id = $item->{ 'Track ID' };
        $new_items[] = $tracks[ $track_id ];
      }

      $playlist->{ 'Playlist Items' } = $new_items;
    }
  }

  //
  // IMPLEMENTATION
  //
  protected function openFileOrSource( $file = NULL, $source = NULL) {

    // Open the XML document in the DOM
    $dom = new DomDocument();

    if ( NULL !== $file ) {
      if ( !file_exists( $file ) ) {
        die( 'iTunes XML file not found: ' . $file );
      }
      if ( !$dom->load( $file ) ) {
        die( 'Could not parse iTunes XML file: ' . $file );
      }
    }
    else if ( NULL !== $source ) {
      if ( !$dom->loadXML( $source ) ) {
        die( 'Could not parse XML source: ' . $source );
      }
    }

    // Get the root element <plist>
    $plist_node = $dom->documentElement;
    $first_dict_node = NULL;

    // First <dict> contains version-info + tracks-node
    foreach ( $plist_node->childNodes as $child ) {
      if ( 'dict' === $child->nodeName ) {
        $first_dict_node = $child;
        break;
      }
    }

    // Fell-through: Parse
    $this->file_name = $file;
    $this->data = $this->parseDict( $first_dict_node, NULL );

  }

  // To be used with the uasort() array function
  protected function sort( $left, $right ) {

    $field = $this->sort_field;
    $direction = $this->sort_direction;

    if ( !isset( $left->{ $field } ) ) {
      return 1;
    }
    elseif ( !isset( $right->{ $field } ) ) {
      return -1;
    }

    // Return the strcmp() of the two fields
    $left = $left->{ $field };
    $right = $right->{ $field };

    switch ( gettype( $left ) ) {

      case 'boolean':
        $left = (int) $left;
        $right = (int) $right;

      case 'integer':
      case 'double':
        if ( 'descending' === $direction ) {
          return $left === $right ? 0 : ( $left > $right ? -1 : 1 );
        }
        else {
          return $left === $right ? 0 : ( $right > $left ? -1 : 1 );
        }
      break;

      default:

        // Detect dates (ISO8601 based), convert to timestamps for comparison
        $rx_date = '/^\d{4}\-\d{2}\-\d{2}T\d{2}\:\d{2}\:\d{2}(Z|\+\d{2}\:\d{2})$/';
        if ( preg_match( $rx_date, $left ) && preg_match( $rx_date, $right ) ) {

          $left = strtotime( $left );
          $right = strtotime( $right );

          if ( 'descending' === $direction ) {
            return $left === $right ? 0 : ( $left > $right ? -1 : 1 );
          }
          else {
            return $left === $right ? 0 : ( $right > $left ? -1 : 1 );
          }

        }

        // Default to a string comparison
        else {
          if ( 'descending' === $direction ) {
            return strcasecmp( $left, $right );
          }
          else {
            return strcasecmp( $right, $left );
          }
        }

    }
  }

  protected function parseDict( $baseNode ) {

    $dicts = array();
    $current_key = NULL;
    $current_value = NULL;

    foreach ( $baseNode->childNodes as $child ) {

      $dict = NULL;

      switch ( $child->nodeName ) {

        case '#text':
        break;

        case 'key':
          $current_key = $child->textContent;
          $current_value = NULL;
        break;

        case 'array':
          $current_value = $this->parseDict( $child );
        break;

        case 'dict':
          $current_value = (object) $this->parseDict( $child );
        break;

        case 'true':
        case 'false':
          $current_value = 'true' === $child->nodeName;
        break;

        case 'integer':
          $current_value = (int) $child->textContent;
        break;

        default:
          $current_value = $child->textContent;

          if ( preg_match( '/^(Music Folder|Location)$/', $current_key ) ) {
            $current_value = urldecode( stripslashes( $current_value ) );
          }

      }

      if ( NULL !== $current_value ) {

        if ( 'array' === $baseNode->nodeName ) {
          $dicts[] = $current_value;
        }
        else if ( NULL !== $current_key ) {
          $dicts[ $current_key ] = $current_value;
          $current_key = NULL;
        }

        $current_value = NULL;

      }

    }

    // Sort the tracks
    if ( $this->sort_field ) {
      uasort( $dicts, array( $this, 'sort' ) );
    }

    return $dicts;

  }

}
